static-checks/ansible/allowed-modules/test.py

```diff
@@ -0,0 +1,127 @@
+    """
+    exprs = set()
+    exprs.add(re.compile(r'<.*Rule.*id="'))
+    exprs.add(re.compile(r'<.*select.*idref="'))
+
+    with open(datastream) as ds:
+        with tempfile.NamedTemporaryFile(mode='w', suffix='.xml') as new_ds:
+            for line in ds:
+                if any(x.search(line) for x in exprs):
+                    line = line.replace('selected="false"', 'selected="true"')
+                new_ds.write(line)
+            new_ds.flush()
+            util.log(f"Saving modified {datastream} with all rules selected as {new_ds.name}")
+            yield Path(new_ds.name)
+
+
+def traverse_blocks(task, allowed_modules):
+    block_set = set(['block'])
+    task_keywords = set(task.keys())
+
+    if len(task_keywords.intersection(block_set)) > 0:
+        for block_task in task['block']:
+            traverse_blocks(block_task, allowed_modules)
+    else:
+        updated_keywords = set()
+        for keyword in task_keywords:
+            keyword = keyword.replace('ansible.builtin.', '')
+            updated_keywords.add(keyword)
+        task_module = list(updated_keywords.intersection(allowed_modules))
+        if len(task_module) == 1:
+            return str(task_module[0])
+        else:
+            module = ''.join(
+                [x for x in list(updated_keywords) if x not in ansible_reserved_keywords]
+            )
+            results.report('fail', module, 'module is not allowed')
+            util.log(f"This occurred in the following task:\n{task}\n")
+            return None
+
+
+def check_allowed_modules(playbook_path, allowed_modules):
+    """
+    Collects all modules from Ansible playbook playbook_path and reports pass
+    if a module is found in the allowed_modules set, otherwise reports fail
+    and prints an offending Ansible module and task. Function also handles
+    Ansible block sections which can bundle multiple tasks.
+    Note: ansible playbook must be a valid playbook which means that
+    'ansible-playbook --syntax-check' must not report any errors for it.
+    """
+    with open(playbook_path) as yamlfile:
+        playbook = yaml.safe_load(yamlfile)
+    modules_list = []
+
+    util.log(f"Check Ansible playbook {playbook_path} contains only allowed modules")
+    for section in playbook:
+        if section['tasks'] is None:
+            # Playbook has no tasks.
+            continue
+        for task in section['tasks']:
+            module = traverse_blocks(task, allowed_modules)
+            if module and modules_list.count(module) == 0:
+                # Module module is not yet in the list, append it.
+                modules_list.append(module)
+
+    for m in sorted(modules_list):
+        results.report('pass', m, 'module is allowed')
+
+
+# rhc-worker-playbook modules, exported per official instructions on
+# https://access.redhat.com/articles/remediation
+os.environ['ANSIBLE_COLLECTIONS_PATH'] = \
+    '/usr/share/rhc-worker-playbook/ansible/collections/ansible_collections/'
```

Comment:
This is RHEL-specific. For Fedora/CentOS we need https://github.com/RHSecurityCompliance/contest/blob/main/hardening/host-os/ansible/test.py#L26 Probably as lib because it will be needed on more places.

Not a big deal now, but mentioning because of inconsistency - `get_all_ansible_modules.sh` is ready for Fedora/CentOS and this part not.

Reply:
I think that's fine. This test is clearly heavily based on something existing, and our other ansible tests are RHEL-only too (the CentOS parts are commented out for now).

Reply:
We should address this for all `contest` tests, so I will leave it as is for now.